### PR TITLE
Fix Dataset/DataArray.isel with drop=True and scalar DataArray indexes

### LINKED ISSUE #6554: isel with drop=True does not drop coordinates if using scalar DataArray as indexer
### What happened?

When using `DataArray/Dataset.isel` with `drop=True` with a scalar DataArray as indexer (see example) resulting scalar coordinates do not get dropped.
When using an integer the behavior is as expected.

### What did you expect to happen?

I expect that using a scalar DataArray behaves the same as an integer.

### Minimal Complete Verifiable Example

```Python
import xarray as xr

da = xr.DataArray([1, 2, 3], dims="x", coord={"k": ("x", [0, 1, 2])})
# <xarray.DataArray (x: 3)>
# array([1, 2, 3])
# Coordinates:
#     k        (x) int32 0 1 2

da.isel({"x": 1}, drop=True)
# works
# <xarray.DataArray ()>
# array(2)

da.isel({"x": xr.DataArray(1)}, drop=True)
# does not drop "k" coordinate
# <xarray.DataArray ()>
# array(2)
# Coordinates:
#     k        int32 1
```

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: 4fbca23a9fd8458ec8f917dd0e54656925503e90
python: 3.9.6 | packaged by conda-forge | (default, Jul  6 2021, 08:46:02) [MSC v.1916 64 bit (AMD64)]
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: AMD64 Family 23 Model 113 Stepping 0, AuthenticAMD
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('de_DE', 'cp1252')
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.18.2.dev76+g3a7e7ca2.d20210706
pandas: 1.3.0
numpy: 1.21.0
scipy: 1.7.0
netCDF4: 1.5.6
pydap: installed
h5netcdf: 0.11.0
h5py: 3.3.0
Nio: None
zarr: 2.8.3
cftime: 1.5.0
nc_time_axis: 1.3.1
PseudoNetCDF: installed
cfgrib: None
iris: 2.4.0
bottleneck: 1.3.2
dask: 2021.06.2
distributed: 2021.06.2
matplotlib: 3.4.2
cartopy: 0.19.0.post1
seaborn: 0.11.1
numbagg: 0.2.1
fsspec: 2021.06.1
cupy: None
pint: 0.17
sparse: 0.12.0
setuptools: 49.6.0.post20210108
pip: 21.3.1
conda: None
pytest: 6.2.4
IPython: None
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by headtr1ck**
- [x] Closes #6554
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Additionally I have added new literal types for error handling (Only applied to functions related to isel such that mypy stops complaining).

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -2182,14 +2182,14 @@ def isel(
...
-        drop : bool, optional
+        drop : bool, default False
             If ``drop=True``, drop coordinates variables indexed by integers
             instead of making them scalar.
         missing_dims : {"raise", "warn", "ignore"}, default: "raise"
             What to do if dimensions that should be selected from are not present in the
             Dataset:
             - "raise": raise an exception
-            - "warning": raise a warning, and ignore the missing dimensions
+            - "warn": raise a warning, and ignore the missing dimensions
```
Comment (max-sixty) [inline xarray/core/dataset.py:2192]: Thanks, nice spot

File: xarray/core/types.py [hunk 2]
```diff
@@ -33,3 +33,6 @@
 DaCompatible = Union["DataArray", "Variable", "DataArrayGroupBy", "ScalarOrArray"]
 VarCompatible = Union["Variable", "ScalarOrArray"]
 GroupByIncompatible = Union["Variable", "GroupBy"]
+
+ErrorChoice = Literal["raise", "ignore"]
```
Comment (max-sixty) [inline xarray/core/types.py:37]: Great! We could find / replace for these at  some point

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -2182,14 +2182,14 @@ def isel(
             If DataArrays are passed as indexers, xarray-style indexing will be
             carried out. See :ref:`indexing` for the details.
             One of indexers or indexers_kwargs must be provided.
-        drop : bool, optional
+        drop : bool, default False
```
Comment (max-sixty) [inline xarray/core/dataset.py:2185]: Weirdly, this doesn't actually get used in `_isel_fancy`...

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -2270,6 +2270,8 @@ def _isel_fancy(
                 }
                 if var_indexers:
                     new_var = var.isel(indexers=var_indexers)
+                    if drop and new_var.ndim == 0:
```
Comment (max-sixty) [inline xarray/core/dataset.py:2273]: Could we link back to this issue (or explain inline, but linking back probably easier)?

**Review by max-sixty — APPROVED** [review]
Excellent, thanks @headtr1ck !

File: xarray/core/dataset.py (hunk 3 above)
Comment (max-sixty) [reply xarray/core/dataset.py:2185]: https://github.com/pydata/xarray/pull/6580

Comment (headtr1ck) [reply xarray/core/dataset.py:2185]: Isn't that destroying the purpose of this PR?

Comment (max-sixty) [reply xarray/core/dataset.py:2185]: It gets used just below! Atm `isel_fancy` doesn't do anything with `drop` — unless I'm missing something (very possible)

Comment (max-sixty) [reply xarray/core/dataset.py:2185]: `drop` gets used here: https://github.com/pydata/xarray/pull/6580/files#diff-763e3002fd954d544b05858d8d138b828b66b6a2a0ae3cd58d2040a652f14638R2236

File: xarray/core/types.py (hunk 2 above)
Comment (headtr1ck) [reply xarray/core/types.py:37]: Even though it is scope creep, I have typed any occurences that I have found in this PR.
Mypy is not complaining locally about these changes.

**Comment by max-sixty** [comment]
I think the error is unrelated — trying again.

Thanks for switching all those over @headtr1ck !

**Pull request merged**